### ISSUE TITLE
PPI with ClusterData & ClusterLoader

### ISSUE THREAD

**Issue by CFF-Dream** [rusty1s/pytorch_geometric#1291]
## ❓ Questions & Help

I processed ppi dataset with ClusterData and CluasterLoader, and set the parameters according to the paper '' Cluster-GCN: An Efficient Algorithm for Training Deep and Large Graph Convolution Networks '', this paper gave a 99.36% accuracy. But my train、val and test accuracy always be 0.000 like 
`Epoch: xx, Loss: 1.4540, Train: 0.0000, Val: 0.0000, Test: 0.0000`
do you know why it happened?Thank you very much!

this is my code:
```
import torch
import time
import torch.nn as nn
import torch.nn.functional as F
import os.path as osp
from torch_geometric.datasets import PPI
from ppi_cluster import ClusterData, ClusterLoader
from torch_geometric.nn import SAGEConv, ChebConv
from sklearn.metrics import f1_score

path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'PPI')
dataset = PPI(path)
train_dataset = PPI(path, split='train')    #20graphs
val_dataset = PPI(path, split='val')        #2graphs
test_dataset = PPI(path, split='test')      #2graphs

print('Partioning the graph... (this may take a while)')
train_dataset_list = []
val_dataset_list = []
test_dataset_list = []
dataset_list = []
train_dataset_index = test_dataset_index =  val_dataset_index = 0

for data in train_dataset:
    cluster_data = ClusterData(data, 'train', train_dataset_index, num_parts=2, recursive=False,
                           save_dir=dataset.processed_dir)
    loader = ClusterLoader(cluster_data, batch_size=20, shuffle=True,
                       num_workers=0)
    train_dataset_list.append(loader)
    dataset_list.append(loader)
    train_dataset_index += 1
    
for data in test_dataset:
    cluster_data = ClusterData(data, 'test', test_dataset_index, num_parts=2, recursive=False,
                           save_dir=dataset.processed_dir)
    loader = ClusterLoader(cluster_data, batch_size=20, shuffle=True,
                       num_workers=0)
    test_dataset_list.append(loader)
    dataset_list.append(loader)
    test_dataset_index += 1

for data in val_dataset:
    cluster_data = ClusterData(data, 'val', val_dataset_index, num_parts=2, recursive=False,
                           save_dir=dataset.processed_dir)
    loader = ClusterLoader(cluster_data, batch_size=20, shuffle=True,
                       num_workers=0)
    val_dataset_list.append(loader)
    dataset_list.append(loader)
    val_dataset_index += 1

print('Done!')

class Net(torch.nn.Module):
    def __init__(self, in_channels, out_channels):
        super(Net, self).__init__()
        dim = 512
        self.gcn1 = ChebConv(in_channels, dim, K=1)
        self.lin1 = nn.Linear(in_channels, dim)
        self.bn1 = nn.BatchNorm1d(dim)
        self.gcn2 = ChebConv(dim, out_channels, K=1)
        self.lin2 = nn.Linear(dim, out_channels)
        self.bn2 = nn.BatchNorm1d(out_channels)
        
    def forward(self, x, edge_index):
        x = F.relu(self.gcn1(x, edge_index) + self.lin1(x))
        x = self.bn1(x)
        x = F.relu(self.gcn2(x, edge_index) + self.lin2(x))
        x = self.bn2(x)
        return F.log_softmax(x, dim=1)

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = Net(dataset.num_features, dataset.num_classes).to(device)
loss_op = torch.nn.BCEWithLogitsLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.05)

def train():
    model.train()
    total_loss = 0
    for loader in train_dataset_list:
        for data in loader:
            data = data.to(device)
            optimizer.zero_grad()
            loss = loss_op(model(data.x, data.edge_index), data.y)
            total_loss += float(loss.item())
            loss.backward()
            optimizer.step()
    return total_loss / len(train_dataset_list)

def test(list):
    model.eval()
    ys, preds, acc = [], [] , 0
    for loader in list:
        for data in loader:
            ys.append(data.y)
            with torch.no_grad():
                out = model(data.x.to(device), data.edge_index.to(device))
            preds.append((out>0).float().cpu())
        y, pred = torch.cat(ys, dim=0).numpy(), torch.cat(preds, dim=0).numpy()
        f1 = f1_score(y, pred, average='micro') if pred.sum() > 0 else 0
        acc += f1
    return acc / len(list)

for epoch in range(1, 1000):
    loss = train()
    train_accs = test(train_dataset_list)
    val_accs = test(val_dataset_list)
    test_accs = test(test_dataset_list)
    print('Epoch: {:02d}, Loss: {:.4f}, Train: {:.4f}, Val: {:.4f}, Test: {:.4f}'.format(epoch, loss, train_accs, val_accs, test_accs))
```

**Comment by rusty1s** [comment]
I guess
```
x = F.relu()
x = self.bn2(x)
return F.log_softmax(x, dim=1)
```
breaks a lot of things. Please try to remove `relu` and batch norm calls before using `softmax`.

Otherwise, the usage of multiple `ClusterLoader` instances look kinda strange to me. You may want to pass `Batch.from_data_list(dataset)` to a single `ClusterLoader` and use it like in a single big graph.

**Comment by CFF-Dream** [comment]
Thanks for your reply!  I have already merged the PPI dataset into a big graph according to your advice, and i modified the code, the train、test accuracy  have be non-zero values . But my code still works very bad. I think theres is any wrong in my code  of calculating loss and accuracy. Do you know how fix it? Thanks again.
```
import os
import time
import torch
import numpy as np
import torch.nn as nn
import os.path as osp
import torch.nn.functional as F
from sklearn.metrics import f1_score 
import torch_geometric.transforms as T
from torch_geometric.nn import ChebConv
from torch_geometric.datasets import PPI
from torch_geometric.datasets import Planetoid
from cluster import ClusterData, ClusterLoader
from torch_geometric.data import DataLoader, Data

path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'PPI')
train_dataset = PPI(path, split='train')    #20graphs
val_dataset = PPI(path, split='val')        #2graphs
test_dataset = PPI(path, split='test')      #2graphs

nodes = 56944
train_nodes = int(np.around(0.8*nodes))   
test_nodes = int(nodes - train_nodes)     
true_mask = torch.tensor([True]).repeat(1, train_nodes).squeeze()
false_mask = torch.tensor([False]).repeat(1, test_nodes).squeeze()
mask = torch.cat((true_mask,false_mask), 0).numpy()
np.random.shuffle(mask)
train_mask = torch.from_numpy(mask)
test_mask = ~train_mask

whole_dataset = Data(edge_index=torch.LongTensor(2,0), x=torch.Tensor(0,50), y=torch.Tensor(0,121), 
                     train_mask=torch.Tensor(56944), test_mask=torch.Tensor(56944))
whole_dataset.train_mask = train_mask
whole_dataset.test_mask = test_mask

for dataset in [train_dataset, val_dataset, test_dataset]:
    for data in dataset:
        whole_dataset.edge_index = torch.cat((whole_dataset.edge_index, data.edge_index), 1)   
        whole_dataset.x = torch.cat((whole_dataset.x, data.x), 0)                               
        whole_dataset.y = torch.cat((whole_dataset.y, data.y), 0)                             
        
print('Partioning the graph... (this may take a while)')
cluster_data = ClusterData(whole_dataset, num_parts=50, recursive=False,
                           save_dir=dataset.processed_dir)
loader = ClusterLoader(cluster_data, batch_size=1, shuffle=True,
                       num_workers=0)                                        
print('Done!')

class Net(torch.nn.Module):
    def __init__(self, in_channels, out_channels):
        super(Net, self).__init__()
        dim = 512
        self.gcn1 = ChebConv(in_channels, dim, K=1)
        self.lin1 = nn.Linear(in_channels, dim)
        self.gcn2 = ChebConv(dim, dim, K=1)
        self.lin2 = nn.Linear(dim, dim)
        self.gcn3 = ChebConv(dim, dim, K=1)
        self.lin3 = nn.Linear(dim, dim)
        self.gcn4 = ChebConv(dim, dim, K=1)
        self.lin4 = nn.Linear(dim, dim)
        self.gcn5 = ChebConv(dim, out_channels, K=1)
        self.lin5 = nn.Linear(dim, out_channels)
        
    def forward(self, x, edge_index):
        x= F.dropout(x, p=0.2, training=self.training)
        x = F.relu(self.gcn1(x, edge_index) + self.lin1(x))
        x= F.dropout(x, p=0.2, training=self.training)
        x = F.relu(self.gcn2(x, edge_index) + self.lin2(x))
        x= F.dropout(x, p=0.2, training=self.training)
        x = F.relu(self.gcn3(x, edge_index) + self.lin3(x))  
        x= F.dropout(x, p=0.2, training=self.training)
        x = F.relu(self.gcn4(x, edge_index) + self.lin4(x))
        x = F.dropout(x, p=0.2, training=self.training)
        x = self.gcn5(x, edge_index) + self.lin5(x)
        return x

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = Net(dataset.num_features, dataset.num_classes).to(device)
loss_op = torch.nn.BCEWithLogitsLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.005)

def train():
    model.train()
    total_loss = total_nodes = 0
    for data in loader:
        data = data.to(device)
        optimizer.zero_grad()
        out = model(data.x, data.edge_index)
        loss = loss_op(out[data.train_mask], data.y[data.train_mask])
        loss.backward()
        optimizer.step()        
        nodes = data.train_mask.sum().item()
        total_loss += loss.item() * nodes
        total_nodes += nodes
    return total_loss / total_nodes     
          
def test():
    model.eval()
    ys, preds = [[], []], [[] ,[]]
    for data in loader:
        data = data.to(device)
        with torch.no_grad():
            out = model(data.x, data.edge_index)
        masks = [data.train_mask, data.test_mask]
        for i, mask in enumerate(masks):
            ys[i].append(data.y)
            preds[i].append((out>0).float().cpu())
    y, pred = torch.cat(ys[0], dim=0).cpu(), torch.cat(preds[0], dim=0).cpu()
    train_acc = f1_score(y, pred, average='micro') if pred.sum() > 0 else 0
    y, pred = torch.cat(ys[1], dim=0).cpu(), torch.cat(preds[1], dim=0).cpu()
    test_acc = f1_score(y, pred, average='micro') if pred.sum() > 0 else 0
    return train_acc, test_acc
    
for epoch in range(1, 500):
    loss = train()     
    train_acc, test_acc = test()
    print('Epoch: {:02d}, Loss: {:.4f}, Train: {:.4f}, Test: {:.4f}, '
          .format(epoch, loss, train_acc, test_acc))
```

**Comment by rusty1s** [comment]
You are not creating your data correctly:
```python
train_dataset = PPI(path, split='train')
train_data_list = [data for data in train_dataset]
for data in train_data_list:
    data.train_mask = torch.ones(data.num_nodes, dtype=torch.bool)
    data.val_mask = torch.zeros(data.num_nodes, dtype=torch.bool)
    data.test_mask = torch.zeros(data.num_nodes, dtype=torch.bool)

# Repeat for `val_data_list` and `test_data_list`...

data = Batch.from_data_list(train_data_list + val_data_list + test_dataset)

cluster_data = ClusterData(data, num_parts=50, recursive=False,
                           save_dir=dataset.processed_dir)
```

**Comment by CFF-Dream** [comment]
Thanks you very much for your reply! The code did work better, but not so much, the loss value is always around 0.5, and the f1_score also show a bad performance , it really confused me.

**Comment by rusty1s** [comment]
Can you send a PR with the example PPI-ClusterGCN code? I can look into it.

**Comment by CFF-Dream** [comment]
Ok, Thank you very much! I have already sent a PR "Update cluster_gcn.py".
